src/covariance_parameters/cholesky.jl

```diff
@@ -0,0 +1,222 @@
+
+"""
+    rsqrt(C::Cholesky)
+
+Computes the right-square root of C. 
+"""
+rsqrt(C::Cholesky) = C.uplo == 'U' ? C.U : adjoint(C.L)
+
+"""
+    lsqrt(C::Cholesky)
+
+Computes the left-square root of C. 
+"""
+lsqrt(C::Cholesky) = adjoint(rsqrt(C))
+
+"""
+    stein(Σ::Cholesky, Φ::AbstractMatrix)
+
+Computes the output of the stein  operator
+
+    Σ ↦ Φ * Σ * Φ'.
+"""
+function stein(Σ::Cholesky, Φ::AbstractMatrix)
+    m, n = size(Φ)
+    work_arr = similar(Φ, n, m)
+
+    mul!(work_arr, rsqrt(Σ), adjoint(Φ))
+    U = positive_qrwoq!(work_arr)
+    Π = utrichol(copy(U))
+    return Π
+end
+
+"""
+    stein(Σ::Cholesky, Φ::Adjoint{<:Number,<:AbstractVector})
+
+Computes the output of the stein  operator
+
+    Σ ↦ Φ * Σ * Φ'.
+
+The return type is a subtype of Real. 
+"""
+function stein(Σ::Cholesky, Φ::Adjoint{<:Number,<:AbstractVector})
+    m, n = size(Φ)
+    work_arr = similar(Φ, n, m)
+
+    mul!(work_arr, rsqrt(Σ), adjoint(Φ))
+    U = positive_qrwoq!(work_arr)
+    Π = abs2(U[1, 1])
+    return Π
+end
+
+"""
+    stein(Σ::Cholesky, Φ::AbstractMatrix, Q::Cholesky)
+
+Computes the output of the stein  operator
+
+    Σ ↦ Φ * Σ * Φ' + Q.
+"""
+function stein(Σ::Cholesky, Φ::AbstractMatrix, Q::Cholesky)
```

Comment:
Why is `Q::Cholesky` required if all it has to satisfy is having a `rsqrt`?

Reply:
This happens a lot below too so if anything should change here it should also below.

Reply:
I am using types to dispatch to the "canonical algorithm". An alternative would of course be to add algorithm types, and then pick algorithm based on the type of algorithm input.

Reply:
Ah ok, I didn't see the canonical algorithm but then of course this makes sense.